dbms/src/Interpreters/Expand.cpp

```diff
@@ -0,0 +1,231 @@
+    if (column.column)
+        column.column = makeNullable(column.column);
+}
+} // namespace
+
+Expand::Expand(const DB::GroupingSets & gss)
+    : group_sets_names(gss)
+{
+    collectNameSet();
+}
+
+String Expand::getGroupingSetsDes() const
+{
+    FmtBuffer buffer;
+    buffer.append("[");
+    for (const auto & grouping_set : group_sets_names)
+    {
+        buffer.append("{");
+        for (const auto & grouping_exprs : grouping_set)
+        {
+            buffer.append("<");
+            buffer.joinStr(grouping_exprs.begin(), grouping_exprs.end());
+            buffer.append(">");
+        }
+        buffer.append("}");
+    }
+    buffer.append("]");
+    return buffer.toString();
+}
+
+/// for cases like: select count(distinct a), count(distinct b) from t;
+/// it will generate 2 group set with <a> and <b>, over which we should
+/// expand one more replica of the source rows from the input block and
+/// identify it with the grouping id in the appended new column.
+///
+/// eg: source block         ==>        replicated block
+///      <a, b>              ==>         <a, b, groupingID>  a new column is appended
+///      1  1       target a -+----->     1  null  groupingID for a =1
+///      2  2                 +----->     2  null  groupingID for b =2
+///                 target b -+----->     null  1  groupingID for a =1
+///                           +----->     null  2  groupingID for b =2
+///
+/// when target a specified group set, other group set columns should be filled
+/// with null value to make group by(a,b) operator to meet the equivalence effect
+/// of group by(a) and group by(b) since the other group set columns has been filled
+/// with null value.
+///
+/// \param input the source block
+/// \return
+
+void Expand::replicateAndFillNull(Block & block) const
+{
+    size_t origin_rows = block.rows();
+    // make a replicate slice, using it to replicate origin rows.
+    std::unique_ptr<IColumn::Offsets> offsets_to_replicate;
+    offsets_to_replicate = std::make_unique<IColumn::Offsets>(origin_rows);
+
+    // get the replicate offset fixed as group set num.
+    IColumn::Offset current_offset = 0;
+    const IColumn::Offset replicate_times_for_one_row = getGroupSetNum();
+
+    // create a column for grouping id.
+    auto grouping_id_column = ColumnUInt64::create();
+    auto & grouping_id_column_data = grouping_id_column->getData();
+    // reserve N times of current block rows size.
+    grouping_id_column_data.reserve(block.rows() * replicate_times_for_one_row);
+
+    for (size_t i = 0; i < origin_rows; ++i)
+    {
+        current_offset += replicate_times_for_one_row;
+        (*offsets_to_replicate)[i] = current_offset;
+
+        // in the same loop, to fill the grouping id.
+        for (UInt64 j = 0; j < replicate_times_for_one_row; ++j)
+        {
+            // start from 1.
+            Field grouping_id = j + 1;
+            grouping_id_column->insert(grouping_id);
+        }
+    }
```

Comment:
use `grouping_id_column_data` directly to avoid virtual function overheads.
```suggestion
    grouping_id_column_data.resize(block.rows() * replicate_times_for_one_row);

    size_t grouping_id_column_index = 0;
    for (size_t i = 0; i < origin_rows; ++i)
    {
        current_offset += replicate_times_for_one_row;
        (*offsets_to_replicate)[i] = current_offset;

        // in the same loop, to fill the grouping id.
        for (UInt64 j = 0; j < replicate_times_for_one_row; ++j)
        {
            // start from 1.
            Field grouping_id = j + 1;
            grouping_id_column_data[grouping_id_column_index] = grouping_id;
        }
    }
```

Reply:
thanks haisheng, learned a lot, got it